src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -1552,7 +1552,7 @@ public static void Sort<T>(T[] array, Comparison<T> comparison)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.comparison);
             }
 
-            ArraySortHelper<T>.Sort(array!, 0, array!.Length, comparison); // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
+            ArraySortHelper<T>.Sort(array!, 0, array!.Length, comparison!); // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
```

Comment:
Is the `!` in `array!.Length` needed?

Reply:
Yes it is needed: shared\System\Array.cs(1555,48): error CS8602: Dereference of a possibly null reference. [E:\repos\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]

Seems like a bug, because the first `!` should tell the compiler that array will not be null on this syntax tree.

Reply:
Ah, I think this is by-design.  The `!` in the first case just suppresses the warning, but the variable isn't actually dereferenced before calling array.Length, so it's still possible that the array.Length would dereference a null.

Reply:
Ah you're right, I think we've hit it earlier and opened an issue and it was by design 😄